### PR TITLE
[Work In Progress] MVP: Hardcoded data will print report

### PR DESCRIPTION
**Pull request by perrymant**
- Hardcoded transactions in the `TransactionMaker`
- `Budget` class handles main budget logic
- `Balance` class handles the current value of the account
- Also added a bunch of idea files

TODO:
- [x] Add documentation in README.md
- [x] Add a formatter for transactions -> strings
- [x] More tests, and practice using TDD

### CODE REVIEW

**Comment by kinbiko** [comment]
Awfully quiet here...

**Comment by perrymant** [comment]
I know, I know...
Unfortunately I've had my partner's nephews here for the last few days (it
turns out kids are terribly hard work - who'd have thunked it!).
Last week I received the Bath Uni reading list - so I started on reading
through some of the material, and have started ploughing through SICP and
have made notes on the first 6 lectures. Have you had a look at it? PDF
<http://www.math.grinnell.edu/~miletijo/m208s15/StructIntCompPrograms.pdf>,
video <https://www.youtube.com/watch?v=2Op3QLzMgSY>. I find it v.
interesting, though perhaps dated and LISP-centric, but hopefully it will
improve my understanding of some core CS ideas.
I have also been working on my music memory game since you suggested making
use of JS for it, and have sorted out the UI in HTML/CSS, with most of the
JS game logic sorted. The thing I'm trying to understand at the moment it
the JS Web Audio API so that I can play back the sounds and add some
tremolo/delay/reverb/EQ effects and have more of a responsive background
ambient soundtrack.
I'll have some time again tomorrow for MoneyMaker™ once these damn kids
have gone (only joking, they're quite fun to have around - even if they
leave a dirty/stick trail everywhere they go - damn them!)
We on for next Wednesday?
Take care,
Thomas

**Review by perrymant — COMMENTED** [review]
My main TODO comments are:
1. is TransactionMaker really the most simple way to get data into the system? I feel that the makeTransactionFromArgs() method should perhaps replace this? However, I'm worried a change here will require many alterations throughout the program.
2. I am unable to print the balance after each transaction - I can only print the final balance.
3. the tests are getting incomprehensible - it suggest the program needs alteration.

Thomas

File: src/io/github/perrymant/moneymaker/BudgetTest.java [hunk 1]
```diff
@@ -0,0 +1,130 @@
...
+        target.updateBalance();
+        assertEquals(-20, balance.get());
+    }
+
+    // GivenATransaction_ReturnReportStringFormat
+
+    @Test
+    public void checkTimeFormatting() {
+        Transaction transaction = new Transaction();
+        assertEquals("2018-08-06", transaction.getTimeNow());
```
Comment (kinbiko) [inline src/io/github/perrymant/moneymaker/BudgetTest.java]: This only passed two days ago, and will never pass again in the future.
If you absolutely need to test time (and I recommend that you don't) then you will have to put error boundaries around it. E.g. today ± 1 day.

File: src/io/github/perrymant/moneymaker/BudgetTest.java [hunk 2]
```diff
@@ -0,0 +1,130 @@
...
+    }
+
+    @Test
+    public void givenTransaction_GetDescription() {
```
Comment (kinbiko) [inline src/io/github/perrymant/moneymaker/BudgetTest.java]: This doesn't test `Budget` so it sounds to me like it doesn't belong here.
Applies to multiple tests in this file.

File: src/io/github/perrymant/moneymaker/BudgetTest.java [hunk 3]
```diff
@@ -0,0 +1,130 @@
...
+        Transaction transaction = new Transaction();
+        transaction.setTransactionType(TransactionType.CREDIT);
+        transaction.setAmount(50);
+        transaction.setDescription("New payment");
+        transaction.setTime("2018/01/02 20:12");
+        assertEquals("2018/01/02 20:12, CREDIT, 50, New payment", transaction.reportStringFormat());
+    }
+
+    @Test
+    public void createMultipleTransactions() {
```
Comment (kinbiko) [inline src/io/github/perrymant/moneymaker/BudgetTest.java]: How is this different logically from `startCanChangeBalanceValue_Credit()`?

File: src/io/github/perrymant/moneymaker/BudgetTest.java [hunk 4]
```diff
@@ -0,0 +1,130 @@
...
+    @Test
+    public void canGenerateStringReportWithUpdatedBalance() {
+        transactions.add(makeTransactionFromArgs("2018/01/02", TransactionType.CREDIT, 150, "Got paid"));
+        transactions.add(makeTransactionFromArgs("2018/01/03", TransactionType.DEBIT, 100, "Got robbed"));
+        Budget budget = new Budget(balance, transactions);
+        budget.updateBalance();
+        assertEquals("2018/01/02, CREDIT, 150, Got paid, 150\n2018/01/03, DEBIT, 100, Got robbed, 50\n", budget.printBalance());
+    }
+
+    private Transaction makeTransactionFromArgs(String dateTime,
```
Comment (kinbiko) [inline src/io/github/perrymant/moneymaker/BudgetTest.java]: Would we lose a lot of confidence in our test if we make some of these parameters constants? E.g. always using the same `description` and `dateTime`?

File: src/io/github/perrymant/moneymaker/Moneymaker.java [hunk 5]
```diff
@@ -2,6 +2,16 @@
...
     }
 }
+
+
+// Main classes should be ~1 line. There is not much need to test it.
+// avoid OR/AND in a method/class names.
+// only 1 public class per file.
+//Paste class code when in the "Project pane" -> auto generate class.
+// POJOs get tested through their use in other classes.
+//MVP: minimal viable product
```
Comment (kinbiko) [inline src/io/github/perrymant/moneymaker/Moneymaker.java]: minimum ;)

File: README.md [hunk 6]
```diff
@@ -1 +1,45 @@
...
+1. TransactionMaker
+    - Hardcode the transactions with makeTransaction()
+    - Uses the Transaction object
+1. Transaction: (a POJO)
+     - String time
+     - TransactionType transactionType: (enum CREDIT/DEBIT)
+     - int amount
+     - String description
+     - String reportStringFormat()
+     Todo: should reportStringFormat() be in it's own class?
```
Comment (kinbiko) [inline README.md:35]: I see where you're coming from here. Some people would put this in another class because they want this class to be as POJO-y as possible. I'm not 100% convinced of the merit of extracting it here.

There's a way of making this a bit more pure. We can override the `toString()` method instead of using this name. All objects have this method anyway and the default implementation doesn't give you much value.

I'll leave the decision up to you:

1. Leave as is
1. Extract a separate class for formatting a transaction as a string.
1. Override `toString()`

File: README.md [hunk 7]
```diff
@@ -1 +1,45 @@
...
+    1. int amount
+    1. int outgoing
+1. display to user
+    1. int balance projection
+
+Classes in use:
+1. Moneymaker (main method)
+    - creates a new Budget using:
+    - new Budget(new Balance(), new TransactionMaker().getTransactions()).printReport();
+    Todo: is TransactionMaker really the most simple way to get data into the system?
```
Comment (kinbiko) [inline README.md:18]: I'm listening...

File: README.md [hunk 8]
```diff
@@ -1 +1,45 @@
...
+     - TransactionType transactionType: (enum CREDIT/DEBIT)
+     - int amount
+     - String description
+     - String reportStringFormat()
+     Todo: should reportStringFormat() be in it's own class?
+
+transaction.reportStringFormat() should look like:
+UTC, transactionType, value, description, balance after transaction.
+
+TODO: unable to print balance after each transaction - I can only print the final balance.
```
Comment (kinbiko) [inline README.md:40]: I think we can figure out where we should put this if we do some debugging.
Alternatively, we can do this in a very clean manner using TDD as well.

File: README.md [hunk 9]
```diff
@@ -1 +1,45 @@
...
+- tried to do this in the Budget.printBalance() for loop with budget.get()...
+- tried to add a budget field in the Transaction class with getters/setters...
+
+Todo: the tests are getting incomprehensible - it suggest the program needs alteration.
+I feel that the makeTransactionFromArgs() method should be part of the program,
+somehow replacing aspects of the TransactionMaker class - but I'm worried a chance here will require
```
Comment (kinbiko) [inline README.md:46]: Yay! Welcome to your first ever legacy codebase! Let's see if we can't gain some more confidence in the correctness of our application before we attempt any dangerous refactoring today!

**Review by kinbiko — COMMENTED** [review]
We can discuss these in a bit in the curfew!

**Review by kinbiko — APPROVED** [review]
LGTM
